src/spec.rs

```diff
@@ -273,3 +277,247 @@ pub(crate) fn path_to_string<T: AsRef<Path>>(path: T) -> Box<str> {
+    where
+        T: AsyncRead + Unpin,
+    {
+        let magic = reader.read_u32_le().await?;
+        if magic != CENTRAL_DIRECTORY_END_SIGNATURE {
+            return Err(ZipError::InvalidArchive("Invalid digital signature header"));
+        }
+        let disk_number = reader.read_u16_le().await?;
+        let disk_with_central_directory = reader.read_u16_le().await?;
+        let number_of_files_on_this_disk = reader.read_u16_le().await?;
+        let number_of_files = reader.read_u16_le().await?;
+        let central_directory_size = reader.read_u32_le().await?;
+        let central_directory_offset = reader.read_u32_le().await?;
+        let zip_file_comment_length = reader.read_u16_le().await? as usize;
+        let mut zip_file_comment = Vec::with_capacity(zip_file_comment_length);
+        reader.read_exact(&mut zip_file_comment).await?;
+        let zip_file_comment: Arc<[u8]> = Arc::from(zip_file_comment);
+        Ok(Self {
+            disk_number,
+            disk_with_central_directory,
+            number_of_files_on_this_disk,
+            number_of_files,
+            central_directory_size,
+            central_directory_offset,
+            zip_file_comment,
+        })
+    }
+
+    pub async fn find_and_parse<T>(reader: &mut T) -> ZipResult<(Self, u64)>
+    where
+        T: AsyncRead + AsyncSeek + Unpin + Send,
+    {
+        const HEADER_SIZE: u64 = 22;
+        const MAX_HEADER_AND_COMMENT_SIZE: u64 = 66000;
+        const BYTES_BETWEEN_MAGIC_AND_COMMENT_SIZE: u64 = HEADER_SIZE - 6;
+        let file_lenght = reader.seek(tokio::io::SeekFrom::End(0)).await?;
+        let search_upper_bound = file_lenght.saturating_add(MAX_HEADER_AND_COMMENT_SIZE);
+        if file_lenght < HEADER_SIZE {
+            return Err(ZipError::InvalidArchive("Invalid zip header"));
+        }
+
+        let mut pos = file_lenght - HEADER_SIZE;
+        while pos >= search_upper_bound {
+            reader.seek(tokio::io::SeekFrom::Start(pos)).await?;
+            if reader.read_u32_le().await? == CENTRAL_DIRECTORY_END_SIGNATURE {
+                reader
+                    .seek(tokio::io::SeekFrom::Current(
+                        BYTES_BETWEEN_MAGIC_AND_COMMENT_SIZE as i64,
+                    ))
+                    .await?;
+                let cde_start_pos = reader.seek(tokio::io::SeekFrom::Start(pos)).await?;
+                if let Ok(end_header) = Self::parse(reader).await {
+                    return Ok((end_header, cde_start_pos));
+                }
+            }
+            pos = match pos.checked_sub(1) {
+                Some(p) => p,
+                None => break,
+            };
+        }
+        Err(ZipError::InvalidArchive(
+            "Could not find central directory end",
+        ))
+    }
+
+    pub async fn write<T: AsyncWrite + Unpin + Send>(&self, writer: &mut T) -> ZipResult<()> {
+        writer.write_u32_le(CENTRAL_DIRECTORY_END_SIGNATURE).await?;
```

Comment:
Just write this to a `Box<[u8]>` and asynchronously write that -- it'll be easier to update once #93 is merged.